File: src/main/java/com/github/bgalek/hackserver/application/challenge/api/ChallengeDefinition.java
package com.github.bgalek.hackserver.application.challenge.api;

import com.github.slugify.Slugify;

import java.util.List;

public interface ChallengeDefinition {

    Slugify slugify = Slugify.builder().build();

    String getName();

    String getDescription();

    String getChallengeEndpoint();

    List<QueryParam> getChallengeParameters();

    Class<?> solutionType();

    TaskDefinition getExample();

    List<TaskDefinition> getTasks();

    default String getId() {
        return slugify.slugify(getName());
    }

    default int getMaxPoints() {
        return this.getTasks()
                .stream()
                .mapToInt(task -> task.getTaskScoring().getMaxPoints())
                .sum();
    }

    record QueryParam(String name, String desc) {}
}

File: src/main/java/com/github/bgalek/hackserver/application/challenge/api/ChallengeDefinition.kt
package com.github.bgalek.hackserver.application.challenge.api

import com.github.slugify.Slugify

object Slugifier {
    private val slug = Slugify.builder().build()

    fun slugify(text: String): String = slug.slugify(text)
}

interface ChallengeDefinition<SOLUTION> {
    val name: String
    val description: String
    val challengeEndpoint: String
    val challengeParameters: List<QueryParam>
    val example: TaskDefinition<SOLUTION>
    val tasks: List<TaskDefinition<SOLUTION>>

    fun getId(): String = Slugifier.slugify(name)
    fun getMaxPoints() = tasks.sumOf { it.score.points }

    @JvmRecord
    data class QueryParam(val name: String, val desc: String)
}
